feat: Add support for listing autolink references

hoffm opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

Fixes #10118.

This is the first `gh repo autolink` subcommand, so this PR adds scaffolding for that command as well. See the linked issue for spec.

hoffm committed:
Introduce repo autolinks list commands

jtmcg commented on pkg/cmd/repo/autolink/autolink.go:
```diff
@@ -0,0 +1,19 @@
+	cmd := &cobra.Command{
+		Use:   "autolink <command>",
+		Short: "Manage autolink references",
+		Long:  "Work with GitHub autolink references.",
```
I think it would be helpful to add some links to relevant documentation. I've added the `heredoc` import and updated the Long doc field to show what I mean:

```suggestion
import (
        "github.com/MakeNowJust/heredoc"
	"github.com/cli/cli/v2/pkg/cmdutil"
	"github.com/spf13/cobra"
)

func NewCmdAutolink(f *cmdutil.Factory) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "autolink <command>",
		Short: "Manage autolink references",
		Long:  heredoc.Docf(`
		Work with GitHub autolink references.
		
		Information about autolinks is only available to repository administrators and can be found at
		https://github.com/{owner}/{repo}/settings/key_links.
		Use %[1]sgh repo autolink list --web%[1]s to open this page for the current repository.
		
		For more information about GitHub autolinks, see https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-autolinks-to-reference-external-resources
	`, "`"),
```

hoffm replied:
Resolved via https://github.com/hoffm/github-cli/pull/1, which is now merged in.

jtmcg commented on pkg/cmd/repo/autolink/http.go:
```diff
@@ -0,0 +1,44 @@
+
+	if resp.StatusCode > 299 {
+		return nil, api.HandleHTTPError(resp)
+	}
```
The AC mentions helpful error messaging for 404's

```
error getting autolinks: HTTP 404: Must have admin rights to Repository. (https://api.github.com/repos/{owner}/{repo}/autolinks)
```

I think that would require an extra conditional, here:

```suggestion
        if resp.StatusCode == 404 {
		return nil, fmt.Errorf("error getting autolinks: HTTP 404: Must have admin rights to Repository. (https://api.github.com/repos/%s)", path)
	} else if resp.StatusCode > 299 {
		return nil, api.HandleHTTPError(resp)
	}
```

Note, this suppresses the actual error returned by the api for 404's. Maybe that's intentional, but I'd look to @andyfeller to chime in, here.

No need to block merging this on his response, though. We can always open a follow-up PR to change the error handling if the AC needs revisiting 🙂

hoffm replied:
This messaging was added via https://github.com/hoffm/github-cli/pull/1. I'll leave the comment unresolved for @andyfeller's benefit.

andyfeller replied:
Interesting 🤔 when I was originally authoring the AC, I was basing this on the response provided back by `gh api` making the necessary calls.

@jtmcg makes a good point that it needs to be easy for users to understand if they don't have the right permission and what they should do to correct it.  In this case, it isn't something trivial like `gh auth refresh -s <SCOPE>` but the role assigned to you.

Let me catch up on the rest of these changes as I trust y'all worked something up while I was OoO.

jtmcg commented on pkg/cmd/repo/autolink/list_test.go:
```diff
@@ -0,0 +1,256 @@
+			cmd.SetErr(&bytes.Buffer{})
+
+			_, err = cmd.ExecuteC()
+			if tt.wantErr {
+				require.EqualError(t, err, tt.errMsg)
+				return
+			}
+
+			require.NoError(t, err)
+			assert.Equal(t, tt.output.WebMode, gotOpts.WebMode)
+			assert.Equal(t, tt.wantExporter, gotOpts.Exporter != nil)
+		})
+	}
```
The `return` here feels weird to me... I think it's a bit easier to read/understand if written with a conditional instead of an early return:

```suggestion
			if tt.wantErr {
				require.EqualError(t, err, tt.errMsg)
			} else {
			        require.NoError(t, err)
			        assert.Equal(t, tt.output.WebMode, gotOpts.WebMode)
			        assert.Equal(t, tt.wantExporter, gotOpts.Exporter != nil)
		        }
		})
	}
```

jtmcg commented on pkg/cmd/repo/autolink/list_test.go:
```diff
@@ -0,0 +1,256 @@
+			name:  "list tty",
+			opts:  &listOptions{},
+			isTTY: true,
+			httpStubs: func(t *testing.T, reg *httpmock.Registry) {
+				reg.Register(
+					httpmock.REST("GET", "repos/OWNER/REPO/autolinks"),
+					httpmock.StringResponse(`[
+						{
+							"id": 1,
+							"key_prefix": "TICKET-",
+							"url_template": "https://example.com/TICKET?query=<num>",
+							"is_alphanumeric": true
+						},
+						{
+							"id": 2,
+							"key_prefix": "STORY-",
+							"url_template": "https://example.com/STORY?id=<num>",
+							"is_alphanumeric": false
+						}
+					]`),
+				)
+			},
```
I _really_ dislike this testing pattern as it results in overmocking and hard to maintain test suites. I think, since this is a new subcommand set, we may have an opportunity to explore a better pattern, here.

I'll open a new branch against this PR to demonstrate what I mean.

jtmcg replied:
Here's my [PR](https://github.com/hoffm/github-cli/pull/1) demonstrating what I'd prefer to see. It can be merged into this branch. Happy to have any discussion you'd like about the implementation pattern there 🙂

hoffm replied:
Thank you for the PR. I agree it's a much better pattern. With my first contribution, I aimed to repeat extant patterns in the codebase, but I'd like to see this pattern propagated throughout.

williammartin replied:
Is a totally sensible decision @hoffm, with all the original maintainers having moved off the team, we're currently figuring out how we would like to address some of these pain points, and as a new command this looks like interesting ground for trying out some of potential patterns.

jtmcg commented on pkg/cmd/repo/autolink/list.go:
```diff
@@ -0,0 +1,123 @@
+package autolink
```
To follow our existing patterns, this should probably go under a `list` package in a subdirectory. That said, I can see the value in not adding that abstraction quite yet, as there's only a single subcommand to autolink being introduced in this PR, and instead adding it in the next PR that introduces the second subcommand.

What do you think?

hoffm replied:
I went back and forth on this myself. I noticed `repo/deploy-key` uses a multi-package structure. Elsewhere (outside of `repo/`), the `label` command uses a flat, single-package structure. Given that ambiguity and the fact that—as you note—there is only one subcommand in this dir for now, I came down on the side of "flat".

However, prompted by your question, I am now noticing that within `repo/`, there are actually _three_ commands with subcommands—`deploy-key`, `gitignore`, and `license`—and they all use subdirectories. For this reason I think it makes sense to introduce the abstraction now. I plan to develop the other subcommands soon (if no one beats me to it), and if for some reason that doesn't happen, I don't want to muddy the waters as to the preferred structure.

hoffm replied:
I implemented this change via 955bd6b2480ac045105a55cd5480fc91b247f93c.

If the `get` wants to reuse the `autolink` struct type, we'll have to move that type out of the `list` package at that point.

jtmcg reviewed (commented):
Thanks for getting started on this! I've begun reviewing this PR by verifying the AC. I've only encountered one discrepancy

### Missing admin rights

> Given I don't have the admin role on the repository
> And Given I have a local repository cloned from GitHub
> When I run `gh repo autolink list`
> Then I see an informative error
> 
> `error getting autolinks: HTTP 404: Must have admin rights to Repository. (https://api.github.com/repos/{owner}/{repo}/autolinks)`

I currently see this error:

```
HTTP 404: Not Found (https://api.github.com/repos/{owner}/{repo}/autolinks)
```

It appears that the api response doesn't specify that you don't have the sufficient permissions and just responds with the `404`. Since you've piped the error directly into the response here, it makes sense that this is the limit of the error message.

The AC asks for a more helpful error message, so we should probably wrap this error in a custom error response. I've made note of where I think this needs to happen in the comments review comments.

Otherwise, all the other AC seems to be working as outlined.

I have added a few other comments for your consideration. There's one in there about the testing pattern this has copied that is one of my personal pain-points in `gh` development... I'm not gonna block this PR on it, but I do plan to open a branch against this one to at least demonstrate how I would do this differently so that we can get away from this over-mocking pattern. 

I'll keep you posted on the status of that.

Overall, excellent work, and thank you for getting after this!

jtmcg committed:
Apply PR comment changes

jtmcg committed:
Refactor autolink list and test to use http interface for simpler testing

This defines an AutolinkClient interface with a Get method used for
fetching the autolinks lists from the api. Then, the http client for
autolinks implements this interface with the AutolinkGetter struct.

This allows for dependency injection of the AutolinkGetter struct into the
listOptions, enabling mocking of the AutolinkGetter for testing. The
result of this is simpler tests that are easier to maintain, because the
interface for the table tests now allow for defining autolink structs as
the response instead of large mocked api calls.

This also allows for bespoke testing of the http file, which I'll follow
up with in the next commit.

jtmcg committed:
Add testing for AutoLinkGetter

This adds the missing mocked http tests to the http_test.go file. These
tests were previously bundled with the tests in list_test.go, creating a
testing pattern that was difficult to understand and maintain. The
refactor in the previous commit replaced these tests with the
AutolinkClient interface, allowing for the httpmocks to be isolated to the
AutolinkGetter that implements that interface.

hoffm committed:
Refactor out early return in test code

Co-authored-by: Tyler McGoffin <jtmcg@github.com>

hoffm committed:
Whitespace

hoffm committed:
Refactor autolink subcommands into their own packages

williammartin commented on pkg/cmd/repo/autolink/list/http.go:
```diff
@@ -0,0 +1,56 @@
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == 404 {
```
**Nit**

Prefer matching against `http.StatusNotFound`

williammartin commented on pkg/cmd/repo/autolink/list/http.go:
```diff
@@ -0,0 +1,56 @@
+	defer resp.Body.Close()
+
+	if resp.StatusCode == 404 {
+		return nil, fmt.Errorf("error getting autolinks: HTTP 404: Must have admin rights to Repository. (https://api.github.com/%s)", path)
```
**Suggestion**

I suspect the right thing to do here to write this error message as a "maybe it's because of this form" i.e.

```
HTTP 404: Perhaps you are missing admin rights to the repository?
```

This indicates the nature of the error, and suggests a possible reason beyond the normal 404.

andyfeller replied:
Yeah, this aspect of GitHub API response codes always make it unclear if it is a permission issue or if it really doesn't exist.  I like the suggested language

williammartin commented on pkg/cmd/repo/autolink/list/list_test.go:
```diff
@@ -0,0 +1,247 @@
+			wantExporter: true,
+		},
+		{
+			name:    "invalid json flag",
+			input:   "--json invalid",
+			wantErr: true,
+			errMsg: heredoc.Doc(`
+				Unknown JSON field: "invalid"
+				Available fields:
+				  id
+				  isAlphanumeric
+				  keyPrefix
+				  urlTemplate`),
+		},
+	}
+
```
**Suggestion**

Instead of this, I'd like to see a test like https://github.com/cli/cli/blob/48bc3cababce65bda8e185f28a467404daf7a76c/pkg/cmd/issue/view/view_test.go#L25-L48

williammartin commented on pkg/cmd/repo/autolink/list/http.go:
```diff
@@ -0,0 +1,56 @@
+	HttpClient *http.Client
+}
+
+func NewAutolinkGetter(httpClient *http.Client) *AutolinkGetter {
```
**Suggestion**

Personally I'd be inclined to remove this since it doesn't really provide any value over the already exposed public `AutolinkGetter` type.

williammartin commented on pkg/cmd/repo/autolink/list/http.go:
```diff
@@ -0,0 +1,56 @@
+
+func (a *AutolinkGetter) Get(repo ghrepo.Interface) ([]autolink, error) {
+	path := fmt.Sprintf("repos/%s/%s/autolinks", repo.RepoOwner(), repo.RepoName())
+	url := ghinstance.RESTPrefix(repo.RepoHost()) + path
```
**Question**

Should we be using `ghrepo.GenerateRepoURL` here, as is done in the command itself?

hoffm replied:
I could definitely be confused about what's going on here, but I don't _think_ we can? `ghinstance.RESTPrefix` supplies the `api` subdomain (or path component in the case of an enterprise instance). This allows us to build the correct URL for the REST endpoint we want to hit here. So it's the difference between:

* Current implementation: `https://api.github.com/repos/OWNER/REPO/autolinks`
* `ghrepo.GenerateRepoURL(repo, "autolinks")`: `https://github.com/OWNER/REPO/autolinks`

Maybe there's some other way to use `GenerateRepoURL` to get the REST endpoint though?

In the command, we use `ghrepo.GenerateRepoURL` to generate the URL for the web interface that displays the autolinks, not an API endpoint.

williammartin replied:
You're correct. Thanks!

williammartin commented on pkg/cmd/repo/autolink/list/http.go:
```diff
@@ -0,0 +1,56 @@
+	}
+}
+
+func (a *AutolinkGetter) Get(repo ghrepo.Interface) ([]autolink, error) {
```
**Suggestion**

I suspect this might be better as `List` rather than `Get` (and same for the struct type). I think it better implies the intention, particularly at the call site in the command.

jtmcg replied:
I like that. Especially since if we continue to copy this pattern throughout the command set the naming for each command's http interface will be consistent 👍

williammartin commented on pkg/cmd/repo/autolink/list/http.go:
```diff
@@ -0,0 +1,56 @@
+		return nil, api.HandleHTTPError(resp)
+	}
+
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var autolinks []autolink
+	err = json.Unmarshal(b, &autolinks)
+	if err != nil {
+		return nil, err
+	}
```
```suggestion	
        var autolinks []autolink
	err = json.NewDecoder(resp.Body).Decode(&autolinks)
	if err != nil {
		return nil, err
	}
```

In practice the difference is cpu and memory impacts are minimal because the response is probably small but I think it's generally more idiomatic to lean on streaming unless there's a good reason not to (like wanting the raw bytes for debugging).

williammartin reviewed (commented):
Overall this is looking really good, thank you for your work.

williammartin commented on pkg/cmd/repo/autolink/list/http.go:
```diff
@@ -0,0 +1,56 @@
+
+	if resp.StatusCode == 404 {
+		return nil, fmt.Errorf("error getting autolinks: HTTP 404: Must have admin rights to Repository. (https://api.github.com/%s)", path)
+	} else if resp.StatusCode > 299 {
```
**Food For Thought**

Not for now but it really feels like this pattern should be pulled into a function. It's used _everywhere_ and I think giving it a name would really help with understanding the intent.

williammartin commented on pkg/cmd/repo/autolink/list/list_test.go:
```diff
@@ -0,0 +1,247 @@
+		response   []autolink
+		wantStdout string
+		wantStderr string
+		wantErr    bool
```
**Suggestion**

I think where possible we should be specific about the error we expect to be returned.

There's already a bug in these tests as a result of being non-specific. The `no results` test **should** be returning a NewNoResultsError but it doesn't because the test block returns an error it `tt.wantErr` is true.

hoffm replied:
Good catch. Fixed via da826db34227bb8aeb8d7e273a9abafcb6e7e748.

williammartin commented on pkg/cmd/repo/autolink/list/list_test.go:
```diff
@@ -0,0 +1,247 @@
+		name       string
+		opts       *listOptions
+		isTTY      bool
+		response   []autolink
```
**Suggestion**

In order to fix https://github.com/cli/cli/pull/10124/files#r1896710006 I'd encourage removing `response`, and the two "mocks" above and instead doing the following:

```go
type stubAutoLinkGetter struct {
	autolinks []autolink
        err error
}

func (g stubAutoLinkGetter) Get(repo ghrepo.Interface) ([]autolink, error) {
	return m.autolinks, m.err
}
```

And in the tests instantiate an instance:

```go
stubGetter: stubAutoLinkGetter{
   ...
}
```

jtmcg replied:
Agreed, that's much better 👍 Not sure why I didn't do that on my first pass 😅

hoffm replied:
Implemented in da826db34227bb8aeb8d7e273a9abafcb6e7e748. Have a look!

hoffm committed:
PR nits

hoffm committed:
Break out autolink list json fields test

hoffm committed:
Remove NewAutolinkClient

hoffm committed:
Use 'list' instead of 'get' for autolink list type and method

hoffm committed:
Decode instead of unmarshal

hoffm committed:
Better error testing for autolink TestListRun

hoffm committed:
Complete get -> list renaming

jtmcg commented on pkg/cmd/repo/autolink/list/list_test.go:
```diff
@@ -0,0 +1,269 @@
+	}
+}
+
+type (
+	testAutolinkClientListError struct{}
+
+	stubAutoLinkLister struct {
+		autolinks []autolink
+		err       error
+	}
+)
+
+func (g stubAutoLinkLister) List(repo ghrepo.Interface) ([]autolink, error) {
+	return g.autolinks, g.err
+}
+
+func (e testAutolinkClientListError) Error() string {
+	return "autolink client list error"
+}
```
Small style nit: I see no other instances of types declared in blocked parens like this. I think we should change this to 

```suggestion
type stubAutoLinkLister struct {
	autolinks []autolink
	err       error
}

func (g stubAutoLinkLister) List(repo ghrepo.Interface) ([]autolink, error) {
	return g.autolinks, g.err
}

type testAutolinkClientListError struct{}

func (e testAutolinkClientListError) Error() string {
	return "autolink client list error"
}
```

I think it also might make sense to move these above `TestListRun` for readability, but I don't feel strongly.

jtmcg reviewed (commented):
This is looking really good! I've got only the one small comment but would otherwise give this my ✅

hoffm committed:
Merge branch 'cli:trunk' into autolink-references

hoffm committed:
Separate type decrarations

jtmcg reviewed (approved):
🔥 Woot! Thanks for all the work to get this one over the line, @hoffm!
